src/table/table-role/use-grid-navigation.ts

```diff
@@ -2,31 +2,42 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { useEffect, useMemo } from 'react';
-import { findFocusinCell, moveFocusBy, restoreTableFocusables, updateTableFocusables } from './utils';
+import {
+  defaultIsSuppressed,
+  findFocusinCell,
+  moveFocusBy,
+  restoreTableFocusables,
+  updateTableFocusables,
+  getFocusables,
+} from './utils';
 import { FocusedCell, GridNavigationAPI, GridNavigationProps } from './interfaces';
 import { KeyCode } from '../../internal/keycode';
 import { nodeContains } from '@cloudscape-design/component-toolkit/dom';
+import { useStableCallback } from '@cloudscape-design/component-toolkit/internal';
 
 /**
- * Makes table with role="grid" navigable with keyboard commands.
+ * Makes table navigable with keyboard commands.
  * See https://www.w3.org/WAI/ARIA/apg/patterns/grid
  */
-export function useGridNavigation({ tableRole, pageSize, getTable }: GridNavigationProps): GridNavigationAPI {
+export function useGridNavigation({
+  active,
+  pageSize,
+  getTable,
+  isSuppressed,
+}: GridNavigationProps): GridNavigationAPI {
   const model = useMemo(() => new GridNavigationModel(), []);
 
+  const getTableStable = useStableCallback(getTable);
+  const isSuppressedStable = useStableCallback((element: HTMLElement) => isSuppressed?.(element) ?? false);
+
   // Initialize the model with the table container assuming it is mounted synchronously and only once.
-  useEffect(
-    () => {
-      if (tableRole === 'grid') {
-        const table = getTable();
-        table && model.init(table);
-      }
-      return () => model.destroy();
-    },
-    // Assuming getTable is stable.
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [model, tableRole]
-  );
+  useEffect(() => {
+    if (active) {
+      const table = getTableStable();
+      table && model.init(table, isSuppressedStable);
+    }
+    return () => model.destroy();
```

Comment:
In this case it is nothing special really. The idea here is to attach a set of behaviours to the table conditionally (when the keyboard nav is used). The behaviours are initialised when the table mounts and the cleanup is performed on unmount. The settings (currently it is only the pageSize) that can be changed dynamically are set from the properties.

This way, the feature is properly encapsulated inside the GridNavigationModel and its binding to the React lifecycle is straightforward (let me know if some of the comments below should be moved to the actual code):

```
function useGridNavigation({ active, pageSize, getTable, isSuppressed }) {
  // Create a stable model reference so that its state is fully controlled with the lifecycle methods below.
  // It is similar to creating a React ref object and then updating it in effects or handlers.
  const model = useMemo(() => new GridNavigationModel(), []);

  // These handlers do not have to be stable but currently it is not expected for those to change.
  // Otherwise, those can be synced with the model same as pageSize.
  const getTableStable = useStableCallback(getTable);
  const isSuppressedStable = useStableCallback((element) => isSuppressed?.(element) ?? false);

  // This effect takes three stable arguments to satisfy the linter. In fact, it only fires when `active` changes.
  // The effect calls the init and cleanup methods on the model instance so that the model lifecycle is in sync with the  table lifecycle pretty much.
  useEffect(() => {
    if (active) {
      const table = getTableStable();
      table && model.init(table, isSuppressedStable);
    }
    return () => model.destroy();
  }, [model, active, getTableStable, isSuppressedStable]);

  // Those properties that can change dynamically can be passed in the same effect that initialises the model or separately. I prefer the second option as there is no need to re-instantiate everything.
  useEffect(() => {
    model.update({ pageSize });
  }, [model, pageSize]);
}
```

The code from above can be represented in a more conventional way but it makes it harder to follow to my liking:

```
function useGridNavigation({ active, pageSize, getTable, isSuppressed }) {
  // The React refs are used to keep whatever state is needed that should not cause the component to re-render.
  const prevFocusedCellRef = useRef(null);
  const focusedCellRef = useRef(null);
  const pageSizeRef = useRef(pageSize);

  // No changes here.
  const getTableStable = useStableCallback(getTable);
  const isSuppressedStable = useStableCallback((element) => isSuppressed?.(element) ?? false);

  // The logic is now encapsulated inside the effect pretty much but it also requires those refs keeping state between renders. (The first two can technically be enclosed inside of the useEffect in this particular case).
  // Same as above, the pageSize can be added to the dependency array which is less efficient but then we don't need a second useEffect.
  useEffect(() => {
    if (active) {
      const table = getTableStable();
      if (!table) {
        return;
      }
      // here goes all code from GridNavigationModel except for the cleanup part
    }
    return () => {
      // the cleanup part goes here
    }
  }, [active, getTableStable, isSuppressedStable]);

  // Sync the properties separately to not cause the init/cleanup to fire unnecessarily.
  useEffect(() => {
    pageSizeRef.current = pageSize;
  }, [pageSize]);
}
```

Reply:
Thank you for the additional explanations! I'd love to know more about the benefits of this pattern (apart from maybe readability), but I think we're gonna do that in a different setting :)

Reply:
It is mainly readability. Another way of achieving the same could be to use a class-based React component where the instance can be used to keep the internal state like element references and observers and the lifecycle methods can be used to update the state with or without issuing a re-render depending on the need. However, the class-based React components are out of fashion and are difficult to test. Besides, those need to be rendered with JSX even if only adding behaviours and the entire rendering is just surpassing the props.children or returning null.

The React hooks are very atomic and it is hard to keep track of refs and states and dependency arrays. Introducing a helper allows encapsulating actual logic while keeping the lifecycle simple. The helper does not have to be a class of course but can be a function with closure state.